src/storage/src/hummock/event_handler/uploader.rs

```diff
@@ -1054,20 +1039,32 @@ impl HummockUploader {
         }
     }
 
-    pub(crate) fn may_flush(&mut self) {
-        if self.context.buffer_tracker.need_more_flush() {
-            self.sealed_data.flush(&self.context, true);
+    pub(crate) fn may_flush(&mut self) -> bool {
+        if !self.context.buffer_tracker.need_more_flush() {
+            return false;
         }
+        self.sealed_data.flush(&self.context, true);
 
         if self.context.buffer_tracker.need_more_flush() {
-            // iterate from older epoch to newer epoch
-            for unsealed_data in self.unsealed_data.values_mut() {
+            let mut unseal_epochs = vec![];
+            for (epoch, unsealed_data) in &self.unsealed_data {
+                unseal_epochs.push((unsealed_data.get_imm_data_size(), *epoch));
+            }
+            // flush large data at first to avoid generate small files.
+            unseal_epochs.sort_by_key(|item| item.0);
+            for (_, epoch) in unseal_epochs.iter().rev() {
+                let task_count =  self.context.stats.uploader_uploading_task_count.get() as u64;
+                if task_count >= self.context.max_upload_task_number {
```

Comment:
I am afraid that the current implementation will introduce a dead lock because `may_flush` will only be triggered when there is either a `BufferMayFlush` or a `ImmToUploader` event, both of which are triggered by `LocalHummockStorage::flush`. Is it possible that all local state store are blocking at waiting for memory and even after `max_upload_task` is triggered and finished, it is still insufficient for any of the local state store to be unblocked? In this case, no more flush can be triggered and we end up in a dead lock.

Can we try whether this patch works good enough without introducing the `max_upload_task_number` limit first? If the goal is to reduce concurrent I/O, I think we can still generate the tasks as usual but have some logics in `UploadTask`, `SyncingData` and `SpillData` to control when to spawn the actual upload task. Improve batching for unsealed epoch is a different story.

Reply:
agree.

Reply:
Ok, I will call this method every time some uploading task finished

Reply:
> Ok, I will call this method every time some uploading task finished

Let's move this to a separate PR and limit the scope of this PR to flush imms out of the epoch order.

Reply:
OK.